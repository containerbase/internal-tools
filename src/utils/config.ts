import is from '@sindresorhus/is';
import { readFile } from '../util';
import type { ConfigFile } from './types';

const keys: (keyof ConfigFile)[] = [
  'datasource',
  'depName',
  'lookupName',
  'buildArg',
  'versioning',
  'latestVersion',
];

function checkArgs(
  cfg: ConfigFile,
  groups: Record<string, string | undefined>
): void {
  for (const key of keys) {
    if (!is.string(cfg[key]) && is.nonEmptyString(groups[key])) {
      cfg[key] = groups[key] as never;
    }
  }
}

export async function readDockerConfig(cfg: ConfigFile): Promise<void> {
  const dockerFileRe = new RegExp(
    '# renovate: datasource=(?<datasource>[a-z-]+?) depName=(?<depName>.+?)(?: lookupName=(?<lookupName>.+?))?(?: versioning=(?<versioning>[a-z-]+?))?\\s' +
      `(?:ENV|ARG) ${cfg.buildArg as string}=(?<latestVersion>.*)\\s`,
    'g'
  );
  const dockerfile = await readFile('Dockerfile');
  const m = dockerFileRe.exec(dockerfile);
  if (m && m.groups) {
    checkArgs(cfg, m.groups);
  }
}
