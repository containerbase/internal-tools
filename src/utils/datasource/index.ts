import { getDatasources } from 'renovate/dist/modules/datasource';
import log from '../logger';

export function register(): void {
  log('register datasources');
  getDatasources();
}
