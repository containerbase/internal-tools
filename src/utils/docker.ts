import is from '@sindresorhus/is';
import * as chalk from 'chalk';
import * as delay from 'delay';
import got, { HTTPError, Headers } from 'got';
import * as wwwAuthenticate from 'www-authenticate';
import { exists } from '../util';
import { cosign } from './cosign/common';
import { docker } from './docker/common';
import log from './logger';
import { ExecError } from './types';

export function splitImage(image: string): {
  registry: string;
  repository: string;
} {
  const parts = image.split('/');
  if (parts.length > 2) {
    return { registry: parts[0], repository: parts.slice(1).join('/') };
  } else {
    return { registry: 'index.docker.io', repository: image };
  }
}

export type DockerManifestV2 = {
  schemaVersion: number;
  config: {
    digest: string;
  };
};

export async function getAuthHeaders(
  registry: string,
  repository: string
): Promise<Headers> {
  try {
    const apiCheckUrl = `https://${registry}/v2/`;
    const apiCheckResponse = await got(apiCheckUrl, { throwHttpErrors: false });
    if (apiCheckResponse.headers['www-authenticate'] === undefined) {
      return {};
    }
    const authenticateHeader = new wwwAuthenticate.parsers.WWW_Authenticate(
      apiCheckResponse.headers['www-authenticate']
    );

    const authUrl = `${authenticateHeader.parms.realm}?service=${authenticateHeader.parms.service}&scope=repository:${repository}:pull`;
    const authResponse = (
      await got<{ token?: string; access_token?: string }>(authUrl, {
        responseType: 'json',
      })
    ).body;

    const token = authResponse.token || authResponse.access_token;
    if (!token) {
      throw new Error('Failed to obtain docker registry token');
    }
    return {
      authorization: `Bearer ${token}`,
    };
  } catch (err) {
    log.error(chalk.red('auth error'), (err as Error).message);
    throw new Error('Failed to obtain docker registry token');
  }
}

export enum DockerContentType {
  ManifestV1 = 'application/vnd.docker.distribution.manifest.v1+json',
  ManifestV1Signed = 'application/vnd.docker.distribution.manifest.v1+prettyjws',
  ManifestV2 = 'application/vnd.docker.distribution.manifest.v2+json',
}

const shaRe = /(sha256:[a-f0-9]{64})/;

export async function getRemoteImageId(
  registry: string,
  repository: string,
  tag = 'latest'
): Promise<string> {
  const headers = await getAuthHeaders(registry, repository);
  headers.accept = DockerContentType.ManifestV2;
  const url = `https://${registry}/v2/${repository}/manifests/${tag}`;

  try {
    const resp = await got<DockerManifestV2>(url, {
      headers,
      responseType: 'json',
    });

    switch (resp.headers['content-type']) {
      case DockerContentType.ManifestV2:
        return resp.body.config.digest;

      case DockerContentType.ManifestV1:
      case DockerContentType.ManifestV1Signed:
        // something wrong, we need to overwrite existing
        log.warn(
          chalk.yellow('Wrong response'),
          `Wrong response: ${resp.headers['content-type'] as string}`
        );
        return '<error>';

      default:
        throw new Error(
          `Unsupported response: ${resp.headers['content-type'] as string}`
        );
    }
  } catch (e) {
    if (e instanceof HTTPError && e.response.statusCode === 404) {
      // no image published yet
      return '<none>';
    }
    log.error(chalk.red('request error'), (e as Error).message);
    throw new Error('Could not find remote image id');
  }
}

export async function getLocalImageId(
  image: string,
  tag = 'latest'
): Promise<string> {
  const res = await docker('inspect', "--format='{{.Id}}'", `${image}:${tag}`);

  const [, id] = shaRe.exec(res.stdout) ?? [];

  if (!id) {
    log.error(res);
    throw new Error('Could not find local image id');
  }

  return id;
}

export type BuildOptions = {
  image: string;
  imagePrefix: string;
  cache?: string;
  cacheTags?: string[];
  tag?: string;
  dryRun?: boolean;
  buildArgs?: string[];
};

const errors = [
  'unexpected status: 400 Bad Request',
  ': no response',
  'error writing layer blob',
];

function canRetry(err: ExecError): boolean {
  return errors.some((str) => err.stderr.includes(str));
}

export async function build({
  image,
  imagePrefix,
  cache,
  cacheTags,
  tag = 'latest',
  dryRun,
  buildArgs,
}: BuildOptions): Promise<void> {
  const args = [
    'buildx',
    'build',
    '--load',
    `--tag=${imagePrefix}/${image}:${tag}`,
  ];

  if (is.nonEmptyArray(buildArgs)) {
    args.push(...buildArgs.map((b) => `--build-arg=${b}`));
  }

  if (is.string(cache)) {
    const cacheImage = `${imagePrefix}/${cache}:${image.replace(/\//g, '-')}`;
    args.push(`--cache-from=${cacheImage}-${tag}`);

    if (is.nonEmptyArray(cacheTags)) {
      for (const ctag of cacheTags) {
        args.push(`--cache-from=${cacheImage}-${ctag}`);
      }
    }

    if (!dryRun) {
      args.push(`--cache-to=type=registry,ref=${cacheImage}-${tag},mode=max`);
    }
  }

  for (let build = 0; ; build++) {
    try {
      await docker(...args, '.');
      break;
    } catch (e) {
      if (e instanceof ExecError && canRetry(e) && build < 2) {
        log.error(chalk.red(`docker build error on try ${build}`), e);
        await delay(5000);
        continue;
      }
      throw e;
    }
  }
}

type PublishOptions = {
  image: string;
  imagePrefix: string;
  tag: string;
  dryRun?: boolean;
};

export async function publish({
  image,
  imagePrefix,
  tag,
  dryRun,
}: PublishOptions): Promise<void> {
  const imageName = `${imagePrefix}/${image}`;
  const fullName = `${imageName}:${tag}`;
  log.info(chalk.blue('Processing image:'), chalk.yellow(fullName));

  log('Fetch new id');
  const newId = await getLocalImageId(imageName, tag);

  log('Fetch old id');
  const { registry, repository } = splitImage(imageName);
  const oldId = await getRemoteImageId(registry, repository, tag);

  if (oldId === newId) {
    log('Image uptodate, no push nessessary:', chalk.yellow(oldId));
    return;
  }

  log('Publish new image', `${oldId} => ${newId}`);
  if (dryRun) {
    log.warn(chalk.yellow('[DRY_RUN]'), chalk.blue('Would push:'), fullName);
  } else {
    await docker('push', fullName);
    await sign(fullName);
  }
  log.info(chalk.blue('Processing image finished:', newId));
}

async function sign(fullName: string): Promise<void> {
  if (!(await exists('cosign'))) {
    log.warn('Cosign is not installed. Skipping container signing');
    return;
  }

  log('Signing image', fullName);
  await cosign('sign', fullName);
}
