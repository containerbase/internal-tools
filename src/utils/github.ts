// istanbul ignore file
import { context, getOctokit } from '@actions/github';
import type { GitHub } from '@actions/github/lib/utils';
import { RequestError } from '@octokit/request-error';
import is from '@sindresorhus/is';
import { getArch, getDistro, readBuffer } from '../util';
import log from './logger';
import type { BinaryBuilderConfig } from './types';

export { getOctokit };

type GitHubOctokit = InstanceType<typeof GitHub>;

interface GhAsset {
  name: string;
}
interface GhRelease {
  id: number;
  name: string | null;
  body?: string | null;

  assets: GhAsset[];

  upload_url: string;
}

let releaseCache: Map<string, GhRelease> | null = null;

export function getBinaryName(
  cfg: BinaryBuilderConfig,
  version: string
): string {
  const arch = getArch();
  if (is.nonEmptyString(arch)) {
    return `${cfg.image}-${version}-${getDistro()}-${arch}.tar.xz`;
  }
  return `${cfg.image}-${version}-${getDistro()}.tar.xz`;
}

function getBody(cfg: BinaryBuilderConfig, version: string): string {
  return `### Bug Fixes

* **deps:** update dependency ${cfg.image} to v${version}`;
}

async function findRelease(
  api: GitHubOctokit,
  version: string
): Promise<GhRelease | null> {
  try {
    if (!releaseCache) {
      const cache = new Map();

      const rels = await api.paginate(api.rest.repos.listReleases, {
        ...context.repo,
        per_page: 100,
      });

      for (const rel of rels) {
        cache.set(rel.tag_name, rel);
      }

      releaseCache = cache;
    }

    return releaseCache.get(version) ?? null;
  } catch (e) {
    if (e instanceof RequestError && e.status !== 404) {
      throw e;
    }
  }
  return null;
}

async function createRelease(
  api: GitHubOctokit,
  cfg: BinaryBuilderConfig,
  version: string
): Promise<GhRelease> {
  try {
    const { data } = await api.rest.repos.createRelease({
      ...context.repo,
      tag_name: version,
      name: version,
      body: getBody(cfg, version),
    });

    releaseCache?.set(data.tag_name, data);
    return data;
  } catch (err) {
    if (
      err instanceof RequestError &&
      err.status == 422 &&
      err.response?.data
    ) {
      log(
        'Release already created by other process, trying to fetch existing:',
        version,
        err.message
      );
      const { data } = await api.rest.repos.getReleaseByTag({
        ...context.repo,
        tag: version,
      });

      releaseCache?.set(data.tag_name, data);
      return data;
    }
    throw err;
  }
}

export async function updateRelease(
  api: GitHubOctokit,
  cfg: BinaryBuilderConfig,
  version: string
): Promise<void> {
  const body = getBody(cfg, version);
  const rel = await findRelease(api, version);
  if (rel == null || (rel.name === version && rel.body === body)) {
    return;
  }
  const { data } = await api.rest.repos.updateRelease({
    ...context.repo,
    release_id: rel.id,
    name: version,
    body,
  });

  releaseCache?.set(data.tag_name, data);
}

export async function uploadAsset(
  api: GitHubOctokit,
  cfg: BinaryBuilderConfig,
  version: string
): Promise<void> {
  try {
    let rel = await findRelease(api, version);
    let release_id = rel?.id ?? 0;

    if (rel == null) {
      rel = await createRelease(api, cfg, version);
      release_id = rel.id;
    }

    const name = getBinaryName(cfg, version);
    const buffer = await readBuffer(`.cache/${name}`);

    const { data } = await api.rest.repos.uploadReleaseAsset({
      ...context.repo,
      release_id,
      url: rel.upload_url,
      // fake because api issues https://github.com/octokit/octokit.js/discussions/2087
      data: buffer as never,
      name,
      headers: {
        'content-type': 'application/octet-stream',
        'content-length': buffer.length,
      },
    });

    // cache asset
    rel.assets.push(data);
  } catch (e) {
    if (e instanceof RequestError && e.status !== 404) {
      throw e;
    }
  }
}

export async function hasAsset(
  api: GitHubOctokit,
  cfg: BinaryBuilderConfig,
  version: string
): Promise<boolean> {
  const rel = await findRelease(api, version);
  const name = getBinaryName(cfg, version);

  return rel?.assets.some((a) => a.name === name) ?? false;
}
