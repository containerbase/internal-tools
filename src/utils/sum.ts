import { BinaryLike, createHash } from 'node:crypto';
import { readBuffer, writeFile } from '../util';
import { getBinaryName } from './config';
import { BinaryBuilderConfig, sumType } from './types';

export function hash(file: BinaryLike): string {
  return createHash(sumType).update(file).digest('hex');
}

export async function createChecksum(
  cfg: BinaryBuilderConfig,
  version: string
): Promise<void> {
  const name = getBinaryName(cfg, version);
  const sumName = getBinaryName(cfg, version, true);
  const buffer = await readBuffer(`.cache/${name}`);

  const sum = hash(buffer);

  await writeFile(`.cache/${sumName}`, sum);
}
