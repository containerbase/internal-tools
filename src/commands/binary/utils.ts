import { getInput } from '@actions/core';
import is from '@sindresorhus/is';
import { getDefaultVersioning } from 'renovate/dist/modules/datasource';
import { getArch, getArg, getDistro, isDryRun, readJson } from '../../util';
import { readDockerConfig } from '../../utils/config';
import {
  DockerPlatform,
  dockerBuildx,
  dockerRun,
} from '../../utils/docker/common';
import log from '../../utils/logger';
import type { BinaryBuilderConfig, ConfigFile } from '../../utils/types';

export async function getConfig(): Promise<BinaryBuilderConfig> {
  const configFile = getInput('config') || 'builder.json';

  const cfg = await readJson<ConfigFile>(configFile);

  if (!is.object(cfg)) {
    throw new Error('missing-config');
  }

  if (!is.string(cfg.image)) {
    cfg.image = getInput('image', { required: true });
  }

  if (!is.string(cfg.buildArg)) {
    cfg.buildArg = cfg.image.toUpperCase() + '_VERSION';
  }

  await readDockerConfig(cfg);

  return {
    ...cfg,
    ignoredVersions: cfg.ignoredVersions ?? [],
    dryRun: isDryRun(),
    lastOnly: getArg('last-only') == 'true',
    buildArgs: getArg('build-args', { multi: true }),
    versioning: cfg.versioning ?? getDefaultVersioning(cfg.datasource),
  } as BinaryBuilderConfig;
}

export async function createBuilderImage(
  ws: string,
  { buildArgs }: BinaryBuilderConfig
): Promise<void> {
  log('Creating builder image');
  const args = [
    'build',
    '--load',
    '-t',
    'builder',
    '--build-arg',
    `DISTRO=${getDistro()}`,
  ];
  const arch = getArch();
  if (is.nonEmptyString(arch)) {
    args.push('--platform', DockerPlatform[arch]);
  }
  if (is.nonEmptyArray(buildArgs)) {
    args.push(...buildArgs.map((b) => `--build-arg=${b}`));
  }
  await dockerBuildx(...args, ws);
}

export async function runBuilder(ws: string, version: string): Promise<void> {
  const args = ['--name', 'builder', '--volume', `${ws}/.cache:/cache`];
  const arch = getArch();
  if (is.nonEmptyString(arch)) {
    args.push('--platform', DockerPlatform[arch]);
  }
  await dockerRun(...args, 'builder', version);
}
