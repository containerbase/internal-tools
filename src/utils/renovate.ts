import * as renovate from './datasource';
import log from './logger';
import * as v from './versioning';

export function register(): void {
  log('register renovate extensions');
  renovate.register();
  v.register();
}
