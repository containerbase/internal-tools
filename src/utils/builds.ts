import is from '@sindresorhus/is';
import {
  ReleaseResult,
  getPkgReleases,
} from 'renovate/dist/modules/datasource';
import { get as getVersioning } from 'renovate/dist/modules/versioning';
import { add as addHostRule } from 'renovate/dist/util/host-rules';
import { configRegexPredicate } from 'renovate/dist/util/regex';
import * as semver from 'semver';
import log from './logger';
import * as renovate from './renovate';

renovate.register();

let latestStable: string | undefined;

export { addHostRule };

function getVersions(versions: string[]): ReleaseResult {
  return {
    releases: versions.map((version) => ({
      version,
    })),
  };
}

export interface BuildsConfig {
  allowedVersions?: string;
  datasource: string;
  depName: string;
  forceUnstable?: boolean;
  ignoredVersions: string[];
  maxVersions?: number;
  lastOnly: boolean;
  latestVersion?: string;
  lookupName?: string;
  startVersion: string;
  versioning: string;
  versions?: string[];
  extractVersion?: string;
}

export interface BuildsResult {
  latestStable: string | undefined;
  versions: string[];
}

export async function getBuildList({
  allowedVersions,
  datasource,
  depName,
  lookupName,
  versioning,
  startVersion,
  ignoredVersions,
  lastOnly,
  forceUnstable,
  versions,
  latestVersion,
  maxVersions,
  extractVersion,
}: BuildsConfig): Promise<BuildsResult | null> {
  log('Looking up versions');
  const ver = getVersioning(versioning);
  const pkgResult = versions
    ? getVersions(versions)
    : await getPkgReleases({
        datasource,
        depName,
        packageName: lookupName,
        versioning,
        extractVersion,
      });
  if (!pkgResult) {
    return null;
  }
  let allVersions = pkgResult.releases
    .map((v) => v.version)
    .filter((v) => ver.isVersion(v) && ver.isCompatible(v, startVersion));

  // filter duplicate versions (16.0.2+7 == 16.0.2+8)
  allVersions = allVersions
    .reverse()
    .filter((v, i) => allVersions.findIndex((f) => ver.equals(f, v)) === i)
    .reverse();

  log(`Found ${allVersions.length} total versions`);
  if (!allVersions.length) {
    return null;
  }
  allVersions = allVersions
    .filter((v) => v === startVersion || ver.isGreaterThan(v, startVersion))
    .filter((v) => !ignoredVersions.includes(v));

  if (!forceUnstable) {
    log('Filter unstable versions');
    allVersions = allVersions.filter((v) => ver.isStable(v));
  }

  if (is.string(allowedVersions)) {
    const isAllowedPred = configRegexPredicate(allowedVersions);
    if (isAllowedPred) {
      allVersions = allVersions.filter((version) => isAllowedPred(version));
    } else if (ver.isValid(allowedVersions)) {
      allVersions = allVersions.filter((version) =>
        ver.matches(version, allowedVersions)
      );
    } else if (semver.validRange(allowedVersions)) {
      allVersions = allVersions.filter((v) =>
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        semver.satisfies(semver.coerce(v)!, allowedVersions)
      );
    } else {
      log.warn(`Invalid 'allowedVersions' options: ${allowedVersions}`);
      return null;
    }
  }

  log(`Found ${allVersions.length} versions within our range`);
  log(`Candidates:`, allVersions.join(', '));

  latestStable =
    latestVersion ||
    /* istanbul ignore next: not testable ts */
    pkgResult.tags?.latest ||
    allVersions.filter((v) => ver.isStable(v)).pop();
  log('Latest stable version is', latestStable);

  if (latestStable && !allVersions.includes(latestStable)) {
    log.warn(
      `LatestStable '${latestStable}' not buildable, candidates:`,
      allVersions.join(', ')
    );
  }

  const lastVersion = allVersions[allVersions.length - 1];
  log('Most recent version is', lastVersion);

  if (is.number(maxVersions) && maxVersions > 0) {
    log(`Building last ${maxVersions} version only`);
    allVersions = allVersions.slice(-maxVersions);
  }

  if (lastOnly) {
    log('Building last version only');
    allVersions = [latestStable && !forceUnstable ? latestStable : lastVersion];
  }

  if (allVersions.length) {
    log('Build list:', allVersions.join(', '));
  } else {
    log('Nothing to build');
  }
  return { versions: allVersions, latestStable };
}
