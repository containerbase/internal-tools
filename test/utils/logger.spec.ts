import * as _core from '@actions/core';
import * as chalk from 'chalk';
import log from '../../src/utils/logger';
import { mocked } from '../utils';

jest.unmock('../../src/utils/logger');

const core = mocked(_core);

describe('utils/logger', () => {
  const logger = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    console.log = logger;
    console.dir = logger;
  });

  it('works', () => {
    log('test');
    log.dir({ name: 'test', code: 1 });
    log.info(chalk.gray('test'), 'it');
    log.warn(chalk.yellow('test'), 'it');
    log.error(chalk.red('test'));

    expect(logger.mock.calls).toMatchSnapshot();
    expect(core.warning).toHaveBeenCalled();
    expect(core.error).toHaveBeenCalled();
  });
});
