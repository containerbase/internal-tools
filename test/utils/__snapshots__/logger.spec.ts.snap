// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils/logger works 1`] = `
[
  [
    "[[34mDEBUG[39m] test",
  ],
  [
    {
      "code": 1,
      "name": "test",
    },
  ],
  [
    "[[32m INFO[39m] [90mtest[39m it",
  ],
  [
    "[[35m WARN[39m] [33mtest[39m it",
  ],
  [
    "[[31mERROR[39m] [31mtest[39m",
  ],
]
`;
